packages/preview-middleware-client/src/adp/quick-actions/common/add-new-annotation-file.ts

```diff
@@ -0,0 +1,94 @@
+import { getDataSourceAnnotationFileMap, writeAnnotationFile } from '../../api-handler';
+import {
+    NESTED_QUICK_ACTION_KIND,
+    NestedQuickAction,
+    NestedQuickActionChild
+} from '@sap-ux-private/control-property-editor-common';
+import { DialogNames, handler } from '../../init-dialogs';
+import OverlayRegistry from 'sap/ui/dt/OverlayRegistry';
+
+export const ADD_NEW_ANNOTATION_FILE = 'add-new-annotation-file';
+
+/**
+ * Add New Annotation File.
+ */
+export class AddNewAnnotationFile implements NestedQuickActionDefinition {
+    public children: NestedQuickActionChild[] = [];
+    readonly kind = NESTED_QUICK_ACTION_KIND;
+    readonly type = ADD_NEW_ANNOTATION_FILE;
+    readonly forceRefreshAfterExecution = true;
+    public get id(): string {
+        return `${this.context.key}-${this.type}`;
+    }
+
+    constructor(protected readonly context: QuickActionContext) {}
+    public get isApplicable(): boolean {
+        return true;
+    }
+    async initialize(): Promise<void> {
+        const dataSourceAnnotationFileMap = await getDataSourceAnnotationFileMap();
+        if (!dataSourceAnnotationFileMap) {
+            throw new Error('No data sources found in the manifest');
+        }
+        for (const key in dataSourceAnnotationFileMap) {
+            const source = dataSourceAnnotationFileMap[key];
+            this.children.push({
+                enabled: true,
+                label: source.annotationDetails.annotationExistsInWS
+                    ? this.context.resourceBundle.getText('SHOW_ANNOTATION_FILE', [key])
+                    : this.context.resourceBundle.getText('ODATA_SORUCE', [key]),
+                children: []
+            });
+        }
+    }
+    async execute(path: string): Promise<FlexCommand[]> {
+        const index = Number(path);
+        if (index >= 0) {
+            // Do not cache the result of getDataSourceAnnotationFileMap api , as annotation file or datasource can be added outside using create command/So refresh would be required for the cache to be updated.
+            const dataSourceAnnotationFileMap = await getDataSourceAnnotationFileMap();
+            const dataSourceId = Object.keys(dataSourceAnnotationFileMap)[index];
+            // Create annotation file only, if no file exists already for datasource id or if the change file exist and but no annotation file exists in file system.
+            if (
+                dataSourceAnnotationFileMap[dataSourceId] &&
+                !dataSourceAnnotationFileMap[dataSourceId].annotationDetails.annotationExistsInWS
+            ) {
+                await writeAnnotationFile({
+                    dataSource: dataSourceId,
+                    serviceUrl: dataSourceAnnotationFileMap[dataSourceId].serviceUrl
+                });
+            } else {
+                const annotationFileDetails = dataSourceAnnotationFileMap[dataSourceId].annotationDetails;
```

Comment:
Is here dataSourceAnnotationFileMap[dataSourceId]  && ....  check needed, like in line 55 ?

Reply:
adapted